controllers/machine_controller_phases.go

```diff
@@ -187,7 +182,22 @@ func (r *MachineReconciler) reconcileBootstrap(ctx context.Context, cluster *clu
 	ready, err := external.IsReady(bootstrapConfig)
 	if err != nil {
 		return err
-	} else if !ready {
+	}
+
+	// Report a summary of current status of the bootstrap object defined for this machine.
+	conditions.SetMirrorCondition(m, clusterv1.BootstrapReadyCondition,
+		conditions.UnstructuredGetter(bootstrapConfig),
+		conditions.WithFallbackValue(ready, clusterv1.WaitingForDataSecretFallbackReason, clusterv1.ConditionSeverityInfo, ""),
+	)
+
+	// If the bootstrap data is populated, set ready and return.
+	if m.Spec.Bootstrap.DataSecretName != nil {
+		m.Status.BootstrapReady = true
```

Comment:
Should this also set clusterv1.BootstrapReadyCondition here as well?

I'm also kind of wondering if the check for `m.Spec.Bootstrap.DataSecretName != nil` should be prioritized above the fetching of the external bootstrap config, since we short circuit here anyway and could never reconcile it if it changes.

Reply:
There is no need to override the mirrored condition because the KubeadmConfig controllers implement its own logic for detecting if the bootstrap data is populated:
https://github.com/kubernetes-sigs/cluster-api/blob/4cbd61f3de68db8113dbaf2921e5b136cd455085/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L224-L230
So we will get ready set on the conditions in tthe KubeadmConfig object and reflected by mirror into BootstrapReadyCondition.

Reply:
What happens if someone is choosing not to use a bootstrap reference, and populate machine.Spec.Bootstrap.DataSecretName manually, which should be possible given the comment here: https://github.com/kubernetes-sigs/cluster-api/blob/0e88a403eb18d26301e2693abfc6611e51712502/api/v1alpha3/machine_types.go#L193-L196

Would we just not set the BootstrapReady condition in that case? If so, what does it look like on the Infrastructure provider side in the conditions world to know it's safe to provision an instance?

Reply:
oh, thanks for the explanation, now I got your point
fixed